std/algorithm/setops.d

```diff
@@ -387,6 +387,16 @@ auto cartesianProduct(RR...)(RR ranges)
                     r = ranges[i].save; // rollover
             }
         }
+        static if (allSatisfy!(hasLength, RR))
+        @property size_t length()
+        {
+            size_t length = 1;
+            foreach (i, ref r; current)
+            {
+                length *= r.length;
```

Comment:
Actually, perhaps if we cache the value of length, this wouldn't be a problem, and it will also solve another problem: `length` as you currently defined it is wrong as soon as you call `popFront`, because it doesn't take into account the fact that part of the generated range may have already been consumed. Caching the value of length would allow us to correctly decrement it in `popFront`, in addition to saving the extra computation.